design-documents/dependencies-injected-in-properties.md

```diff
@@ -0,0 +1,34 @@
+## Introduce the possibility to inject dependencies directly to properties
+### Reasons
+* Allowing backward compatible introduction of new dependencies to classes without having to use ObjectManager directly  
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Currently we have to add optional arguments to classes and invoke ObjectManager::getInstance() to get a dependency
+for cases when the constructor is invoked from a child class. Declaring such dependencies for properties will allow us to avoid this.
+* Allowing circular dependencies  
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;In some cases circular dependencies can be really hard to avoid - by injecting dependencies into properties
+after requested object instances are created we can make circular dependecies possible.
+### Details
+#### XML
+We'd have to add *properties*/*property* elements for *type* elements in di.xml's.
+*property* element would have *type* and *shared* attributes just as *argument* does.  
+#### Injection
+Key points:  
+* Dependencies list for properties will be collected and injected after
```

Comment:
This contradicts with [Magento Technical Guidelines](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
2.2. Object MUST be ready for use after instantiation. No additional public initialization methods are allowed.
2.14. Temporal coupling MUST be avoided

Reply:
No it doesn't. Classes requesting objects from object manager won't get them before they initiated. This is just details regarding implementation of injecting dependencies to properties.

Reply:
To be clear - I'm not suggesting we create setters for these properties or make them public. Objects will be fully initiated after the object manager returns them

Reply:
You cannot create such an object with the "new" operator and use it without additional initialization and we don't have the requirement to create all objects using DI. Actually, we have opposite requirement:  ObjectManager MUST NOT be used in unit tests.